### ISSUE TITLE
partition_data.py hangs

### ISSUE THREAD

**Issue by darioluzzi** [keithf4/pg_partman#83]
Hi,
I'm making some preliminary tests for pg_partman 2.2.2 on postgres 9.4.

I created the following table:

```sql
CREATE TABLE work.matching_input
(
  path_id bigint NOT NULL,
  gps_time timestamp with time zone NOT NULL,
  wgs84_lat integer NOT NULL,
  wgs84_lon integer NOT NULL,
  gps_quality "char" NOT NULL,
  speed_kmh smallint NOT NULL,
  heading smallint NOT NULL,
  write_date timestamp with time zone NOT NULL DEFAULT now(),
  odometer bigint NOT NULL DEFAULT (-1),
  pnt_type "char" NOT NULL DEFAULT 'I'::"char",
  CONSTRAINT matching_input_pk PRIMARY KEY (path_id, gps_time)
)
WITH (
  OIDS=FALSE
);
ALTER TABLE work.matching_input
  OWNER TO partman_user;

CREATE INDEX matching_input_idx
  ON work.matching_input
  USING btree
  (path_id);

CREATE INDEX matching_input_idx2
  ON work.matching_input
  USING btree
  (gps_time);

CREATE TRIGGER matching_input_part_trig
  BEFORE INSERT
  ON work.matching_input
  FOR EACH ROW
  EXECUTE PROCEDURE work.matching_input_part_trig_func();
```

and setup the partitions with partman:
```
    parent_table     |  control   | partition_type | partition_interval |  constraint_cols   | premake | optimize_trigger | optimize_constraint | epoch | inherit_fk | retention  | retention_schema | retention_keep_table | retention_keep_index |  datetime_string  | use_run_maintenance | jobmon | sub_partition_set_full | undo_in_progress 
---------------------+------------+----------------+--------------------+--------------------+---------+------------------+---------------------+-------+------------+------------+------------------+----------------------+----------------------+-------------------+---------------------+--------+------------------------+------------------
 work.matching_input | write_date | time-custom    | 00:05:00           | {path_id,gps_time} |       8 |                4 |                  30 | f     | t          | 35 minutes |                  | f                    | t                    | YYYY_MM_DD_HH24MI | t                   | t      | f                      | f
```
The background worker is running fine as long as I keep inserting data into the table. It keeps creating future tables and dropping the old ones.

The problem originates when I stop inserting data for a time long enough so that no partitions are created and when i resume putting data into the table it gets placed in the main table. 

This is fine, and I'm now triyng to recover missing partitions and split data from the main table to the correct partitions. To do this I'm using the script:
```
./partition_data.py -p 'work.matching_input' -t 'time' -c "host=127.0.0.1 port=5432 dbname=partman user=postgres password=password"  -b 5 -w 5 --autovacuum_on
```

The script never completes though it seems it's active and not stuck waiting:

```
 datid | datname |  pid  | usesysid | usename  | application_name | client_addr | client_hostname | client_port |         backend_start         |          xact_start           |          query_start          |         state_change          | waiting | state  | backend_xid | backend_xmin |                                                           query                                                           
-------+---------+-------+----------+----------+------------------+-------------+-----------------+-------------+-------------------------------+-------------------------------+-------------------------------+-------------------------------+---------+--------+-------------+--------------+---------------------------------------------------------------------------------------------------------------------------
 55900 | partman | 20087 |       10 | postgres |                  | 127.0.0.1   |                 |       50141 | 2016-01-13 10:52:25.403413+01 | 2016-01-13 10:52:25.422262+01 | 2016-01-13 10:52:25.422262+01 | 2016-01-13 10:52:25.422263+01 | f       | active |             |     10065656 | SELECT "partman".partition_data_time('work.matching_input', p_batch_interval := NULL, p_lock_wait := 0, p_order := 'ASC')
```

**Comment by keithf4** [comment]
Working on trying to replicate the issue. One thing I did notice is that you're missing an index on your control column (write_date). If you have a lot of data (which I'm assuming you do if you have to partition in 5 minute intervals), this can have a HUGE impact on performance. Not sure if that's causing the hang, but I would definitely add that.

Will get back to you when I find anything else.

**Comment by keithf4** [comment]
FYI, I have been able to get it to hang while partitioning as well, same as you reported. Will continue to investigate and let you know.

**Comment by darioluzzi** [comment]
I tried adding the index on the control column, but I'm getting the same behavior

**Comment by keithf4** [comment]
Found the error in logic that is causing an infinite loop. This only occurs with custom time partitioning. I'll work on getting a fix out as soon as I can.

**Comment by keithf4** [comment]
I have a fix for this ready for the next version, but am still working on fixing another bug before I release it. Below is the git diff if you need to fix this more immediately.
```
diff --git a/sql/functions/partition_data_time.sql b/sql/functions/partition_data_time.sql
index 52f9e78..4e0098c 100644
--- a/sql/functions/partition_data_time.sql
+++ b/sql/functions/partition_data_time.sql
@@ -101,7 +101,6 @@ FOR i IN 1..p_batch_count LOOP
                 v_min_partition_timestamp := date_trunc('year', v_start_control);
         END CASE;
     ELSIF v_type = 'time-custom' THEN
-        -- Keep going backwards, checking if the time interval encompases the current v_start_control value
         v_time_position := (length(v_last_partition) - position('p_' in reverse(v_last_partition))) + 2;
         v_min_partition_timestamp := to_timestamp(substring(v_last_partition from v_time_position), v_datetime_string);
         v_max_partition_timestamp := v_min_partition_timestamp + v_partition_interval;
@@ -109,9 +108,17 @@ FOR i IN 1..p_batch_count LOOP
             IF v_start_control >= v_min_partition_timestamp AND v_start_control < v_max_partition_timestamp THEN
                 EXIT;
             ELSE
-                v_max_partition_timestamp := v_min_partition_timestamp;
                 BEGIN
-                    v_min_partition_timestamp := v_min_partition_timestamp - v_partition_interval;
+                    IF v_start_control > v_max_partition_timestamp THEN
+                        -- Keep going forward in time, checking if child partition time interval encompasses the current v_start_control value
+                        v_min_partition_timestamp := v_max_partition_timestamp;
+                        v_max_partition_timestamp := v_max_partition_timestamp + v_partition_interval;
+
+                    ELSE
+                        -- Keep going backwards in time, checking if child partition time interval encompasses the current v_start_control value
+                        v_max_partition_timestamp := v_min_partition_timestamp;
+                        v_min_partition_timestamp := v_min_partition_timestamp - v_partition_interval;
+                    END IF;
                 EXCEPTION WHEN datetime_field_overflow THEN
                     RAISE EXCEPTION 'Attempted partition time interval is outside PostgreSQL''s supported time range. 
                         Unable to create partition with interval before timestamp % ', v_min_partition_interval;
@@ -123,6 +130,7 @@ FOR i IN 1..p_batch_count LOOP
 
     v_partition_timestamp := ARRAY[v_min_partition_timestamp];
     IF p_order = 'ASC' THEN
+        -- Ensure batch interval given as parameter doesn't cause maximum to overflow the current partition maximum
         IF (v_start_control + p_batch_interval) >= (v_min_partition_timestamp + v_partition_interval) THEN
             v_max_partition_timestamp := v_min_partition_timestamp + v_partition_interval;
         ELSE
@@ -131,7 +139,7 @@ FOR i IN 1..p_batch_count LOOP
     ELSIF p_order = 'DESC' THEN
         -- Must be greater than max value still in parent table since query below grabs < max
         v_max_partition_timestamp := v_min_partition_timestamp + v_partition_interval;
-        -- Make sure minimum doesn't underflow current partition minimum
+        -- Ensure batch interval given as parameter doesn't cause minimum to underflow current partition minimum
         IF (v_start_control - p_batch_interval) >= v_min_partition_timestamp THEN
             v_min_partition_timestamp = v_start_control - p_batch_interval;
         END IF;
```

**Comment by keithf4** [comment]
I've released version 2.2.3 that incorporates the above fix. Let me know if this works ok for you.
